File: app/schemas/booking.py
from pydantic import BaseModel
from typing import Optional
from pydantic.networks import EmailStr
from app.enums.booking_status import BookingStatus
from datetime import datetime, date
from app.schemas.property import Property


class BookingBase(BaseModel):
    property_id: int
    start_date: date
    end_date: date
    status: Optional[BookingStatus] = BookingStatus.PENDING


class BookingCreate(BookingBase):
    pass


class BookingUpdate(BaseModel):
    status: Optional[BookingStatus] = None
    start_date: Optional[date] = None
    end_date: Optional[date] = None


class Booking(BookingBase):
    id: int
    created_at: datetime
    user_id: int

    class Config:
        orm_mode = True
        from_attributes = True


class PersonalizedOffer(BaseModel):
    property: Property
    discount: float
    message: str

File: frontend/src/types/booking.ts
import { Property } from './api';

export enum BookingStatus {
  PENDING = 'pending',
  CONFIRMED = 'confirmed',
  CANCELLED = 'cancelled'
}

export interface BookingBase {
  property_id: number;
  start_date: string;
  end_date: string;
  status?: BookingStatus;
}

export interface BookingCreate extends BookingBase {}

export interface BookingUpdate {
  status?: BookingStatus;
  start_date?: string;
  end_date?: string;
}

export interface AccessCode {
  id: number;
  booking_id: number;
  code: string;
  valid_from: string;
  valid_until: string;
}

export interface Booking extends BookingBase {
  id: number;
  user_id: number;
  created_at: string;
  property: Property;
  access_code?: AccessCode;
}

export interface PersonalizedOffer {
  property: Property;
  discount: number;
  message: string;
} 